File: libs/moqtail-ts/src/model/control/publish_ok.ts
/**
 * Copyright 2025 The MOQtail Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { BaseByteBuffer, ByteBuffer, FrozenByteBuffer } from '../common/byte_buffer'
import { KeyValuePair } from '../common/pair'
import { Location } from '../common/location'
import { LengthExceedsMaxError } from '../error/error'
import { ControlMessageType, FilterType, GroupOrder } from './constant'
import { MessageParameter, MessageParameters } from '../parameter/message_parameter'
import { Forward } from '../parameter/message/forward'
import { SubscriberPriority } from '../parameter/message/subscriber_priority'
import { GroupOrderParam } from '../parameter/message/group_order_param'
import { SubscriptionFilter } from '../parameter/message/subscription_filter'
import { DeliveryTimeout } from '../parameter/message/delivery_timeout'

export class PublishOk {
  constructor(
    public readonly requestId: bigint,
    public readonly parameters: MessageParameter[],
  ) {}

  getType(): ControlMessageType {
    return ControlMessageType.PublishOk
  }

  serialize(): FrozenByteBuffer {
    const buf = new ByteBuffer()
    buf.putVI(ControlMessageType.PublishOk)
    const payload = new ByteBuffer()
    payload.putVI(this.requestId)

    payload.putVI(this.parameters.length)
    for (const param of this.parameters) {
      payload.putKeyValuePair(param.toKeyValuePair())
    }

    const payloadBytes = payload.toUint8Array()
    if (payloadBytes.length > 0xffff) {
      throw new LengthExceedsMaxError('PublishOk::serialize(payloadBytes.length)', 0xffff, payloadBytes.length)
    }
    buf.putU16(payloadBytes.length)
    buf.putBytes(payloadBytes)
    return buf.freeze()
  }

  static parsePayload(buf: BaseByteBuffer): PublishOk {
    const requestId = buf.getVI()
    const paramCount = buf.getVI()
    const kvps: KeyValuePair[] = []
    for (let i = 0; i < paramCount; i++) {
      kvps.push(buf.getKeyValuePair())
    }
    const parameters = MessageParameters.fromKeyValuePairs(kvps)

    return new PublishOk(requestId, parameters)
  }
}

if (import.meta.vitest) {
  const { describe, test, expect } = import.meta.vitest

  describe('PublishOk', () => {
    test('roundtrip with no parameters', () => {
      const publishOk = new PublishOk(12345n, [])

      const frozen = publishOk.serialize()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.PublishOk))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining)
      const deserialized = PublishOk.parsePayload(frozen)
      expect(deserialized.requestId).toBe(publishOk.requestId)
      expect(deserialized.parameters.length).toBe(0)
      expect(frozen.remaining).toBe(0)
    })

    test('roundtrip with parameters', () => {
      const parameters: MessageParameter[] = [
        new Forward(true),
        new SubscriberPriority(100),
        new GroupOrderParam(GroupOrder.Ascending),
        new SubscriptionFilter(FilterType.LatestObject, undefined, undefined),
      ]
      const publishOk = new PublishOk(12345n, parameters)

      const frozen = publishOk.serialize()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.PublishOk))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining)
      const deserialized = PublishOk.parsePayload(frozen)
      expect(deserialized.requestId).toBe(publishOk.requestId)
      expect(deserialized.parameters.length).toBe(parameters.length)
      expect(frozen.remaining).toBe(0)
    })

    test('roundtrip with AbsoluteRange subscription filter', () => {
      const parameters: MessageParameter[] = [
        new SubscriptionFilter(FilterType.AbsoluteRange, new Location(5n, 10n), 20n),
        new DeliveryTimeout(5000n),
      ]
      const publishOk = new PublishOk(789n, parameters)

      const frozen = publishOk.serialize()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.PublishOk))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining)
      const deserialized = PublishOk.parsePayload(frozen)
      expect(deserialized.requestId).toBe(789n)
      expect(deserialized.parameters.length).toBe(2)
      expect(frozen.remaining).toBe(0)
    })

    test('excess roundtrip', () => {
      const publishOk = new PublishOk(12345n, [new Forward(true)])
      const serialized = publishOk.serialize().toUint8Array()
      const excess = new Uint8Array([9, 1, 1])
      const buf = new ByteBuffer()
      buf.putBytes(serialized)
      buf.putBytes(excess)
      const frozen = buf.freeze()
      const msgType = frozen.getVI()
      expect(msgType).toBe(BigInt(ControlMessageType.PublishOk))
      const msgLength = frozen.getU16()
      expect(msgLength).toBe(frozen.remaining - 3)
      const deserialized = PublishOk.parsePayload(frozen)
      expect(deserialized.requestId).toBe(publishOk.requestId)
      expect(frozen.remaining).toBe(3)
      expect(Array.from(frozen.getBytes(3))).toEqual([9, 1, 1])
    })

    test('partial message', () => {
      const publishOk = new PublishOk(12345n, [new Forward(true)])
      const serialized = publishOk.serialize().toUint8Array()
      const upper = Math.floor(serialized.length / 2)
      const partial = serialized.slice(0, upper)
      const frozen = new FrozenByteBuffer(partial)
      expect(() => {
        frozen.getVI()
        frozen.getU16()
        PublishOk.parsePayload(frozen)
      }).toThrow()
    })
  })
}

File: libs/moqtail-rs/src/model/control/publish_ok.rs
// Copyright 2025 The MOQtail Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use super::constant::ControlMessageType;
use super::control_message::ControlMessageTrait;
use crate::model::common::varint::{BufMutVarIntExt, BufVarIntExt};
use crate::model::error::ParseError;
use crate::model::parameter::message_parameter::{
  MessageParameter, deserialize_message_parameters,
};
use bytes::{BufMut, Bytes, BytesMut};

#[derive(Debug, PartialEq, Clone)]
pub struct PublishOk {
  pub request_id: u64,
  pub parameters: Vec<MessageParameter>,
}

impl PublishOk {
  pub fn new(request_id: u64, parameters: Vec<MessageParameter>) -> Self {
    Self {
      request_id,
      parameters,
    }
  }
}

impl ControlMessageTrait for PublishOk {
  fn serialize(&self) -> Result<Bytes, ParseError> {
    let mut buf = BytesMut::new();
    buf.put_vi(ControlMessageType::PublishOk)?;

    let mut payload = BytesMut::new();
    payload.put_vi(self.request_id)?;

    payload.put_vi(self.parameters.len() as u64)?;
    for param in &self.parameters {
      payload.extend_from_slice(&param.serialize()?);
    }

    let payload_len: u16 = payload
      .len()
      .try_into()
      .map_err(|e: std::num::TryFromIntError| ParseError::CastingError {
        context: "PublishOk::serialize(payload_length)",
        from_type: "usize",
        to_type: "u16",
        details: e.to_string(),
      })?;

    buf.put_u16(payload_len);
    buf.extend_from_slice(&payload);

    Ok(buf.freeze())
  }

  fn parse_payload(payload: &mut Bytes) -> Result<Box<Self>, ParseError> {
    let request_id = payload.get_vi()?;
    let param_count = payload.get_vi()?;
    let parameters =
      deserialize_message_parameters(payload, param_count, ControlMessageType::PublishOk)?;

    Ok(Box::new(PublishOk {
      request_id,
      parameters,
    }))
  }

  fn get_type(&self) -> ControlMessageType {
    ControlMessageType::PublishOk
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::model::common::location::Location;
  use crate::model::control::constant::{FilterType, GroupOrder};
  use bytes::Buf;

  #[test]
  fn test_roundtrip_no_params() {
    let publish_ok = PublishOk::new(123, vec![]);

    let mut buf = publish_ok.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::PublishOk as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let deserialized = PublishOk::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, publish_ok);
    assert!(!buf.has_remaining());
  }

  #[test]
  fn test_roundtrip_with_params() {
    let publish_ok = PublishOk::new(
      456,
      vec![
        MessageParameter::new_forward(true),
        MessageParameter::new_subscriber_priority(5),
        MessageParameter::new_group_order(GroupOrder::Ascending),
        MessageParameter::new_subscription_filter(FilterType::LatestObject, None, None),
      ],
    );

    let mut buf = publish_ok.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::PublishOk as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let deserialized = PublishOk::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, publish_ok);
    assert!(!buf.has_remaining());
  }

  #[test]
  fn test_roundtrip_absolute_range_filter() {
    let publish_ok = PublishOk::new(
      789,
      vec![
        MessageParameter::new_subscription_filter(
          FilterType::AbsoluteRange,
          Some(Location::new(5, 10)),
          Some(100),
        ),
        MessageParameter::new_delivery_timeout(5000),
      ],
    );

    let mut buf = publish_ok.serialize().unwrap();
    let msg_type = buf.get_vi().unwrap();
    assert_eq!(msg_type, ControlMessageType::PublishOk as u64);
    let msg_length = buf.get_u16();
    assert_eq!(msg_length as usize, buf.remaining());
    let deserialized = PublishOk::parse_payload(&mut buf).unwrap();
    assert_eq!(*deserialized, publish_ok);
    assert!(!buf.has_remaining());
  }
}
